### ISSUE TITLE
Hae tietueita 035z-kentällä

### ISSUE THREAD

**Issue by johannaraisa** [KohaSuomi/koha-plugin-broadcast-biblios#14]
Jos tietue saa uuden 035a-kentän arvon, niin vanhat arvot siirtyvät 035z-kenttään. Vastaavuutta pitäisi hakea myös kyseisestä kentästä.

**Comment by johannaraisa** [comment]
Tämä on testattu ja toimii. 

Sain tämän [tietueen](https://tati-test.koha-suomi.fi/cgi-bin/koha/catalogue/detail.pl?biblionumber=2486506) valumaan OUTIn testille. OUTIn testillä oli versio missä oli toinen noista 035z-kentistä 035a-kenttänä.

**Comment by johannaraisa** [comment]
@AnttHeik kokeile sinäkin tehdä testi tämän kanssa.

**Comment by AnttHeik** [comment]
Joo, näyttäisi toimivan hyvin TäTistä OUTIin, mutta törmäsin vielä tällaiseen ongelmaan, kun yritin saada tuotua Melindasta osakohteellisen tietueen. Kun hakee Tietuesiirtäjällä vastaavaa tietuetta, niin vastaan tulee Melindassa deletoitu tietue eli yksi noista aiemmin yhdistetyistä vanhoista tietueista. Kokeilin kahdella eri tietueella: TäTI-id:t 2486884 ja 2486885. Näyttäisi siltä että molemmissa tapauksissa se hakee tietuetta ensimmäisellä 035z-kentän toistumalla eikä 035a-kentän tunnuksella. Liittyyköhän tämä tähän samaan asiaan, vai onkohan tätä tapahtunut jo aiemmin?

**Comment by johannaraisa** [comment]
Kyllä tuo johtuu muutoksesta, että tarkastetaan myös 035z, muokkasin hieman tuota ja minusta näyttää nyt toimivan.

**Comment by AnttHeik** [comment]
Joo, nyt toimi niin kuin piti ja sain osakohteetkin valutettua.

Tekisi mieli vielä testata koko prosessi niin, että Melindassa yhdistettäisiin jokin osakohteellinen tietue ja katsoisi sitten, miten se onnistuu loppuun asti. Tällä hetkellä Melindan Merge-ohjelma ei vain ole toiminnassa, joten en itse pysty sitä testaamaan. Voisin kysyä, onnistuisiko Melindan päässä tehdä tätä jotenkin testillä.

**Comment by johannaraisa** [comment]
Voikohan tätä jo testata Melindan Mergen kanssa?

**Comment by AnttHeik** [comment]
> Voikohan tätä jo testata Melindan Mergen kanssa?

Merge ei ole vieläkään toiminnnassa, mutta meidän piti testata tämä Mintun kanssa toisella tavalla. Se jäi vain joulun aikaan kesken, kun Testi-TäTi ei toiminut. Nyt se on sitten jäänyt muiden kiireiden keskelle, mutta pitää kattoa se nyt tällä viikolla.

**Comment by AnttHeik** [comment]
Tietuesiirtäjä antaa Testi-Tätissä tällaista virheilmoitusta: Request failed with status code 500. 
Onkohan se päällä Melindan suuntaan siellä?

**Comment by johannaraisa** [comment]
> Tietuesiirtäjä antaa Testi-Tätissä tällaista virheilmoitusta: Request failed with status code 500. Onkohan se päällä Melindan suuntaan siellä?

Saattaa olla vielä levällään redusoinnin jäljiltä, katson mistä kiikastaa.

**Comment by johannaraisa** [comment]
Nyt pitäisi toimia koko putki, croni-ajotkin oli näköjään pois päältä.

**Comment by AnttHeik** [comment]
Kiitos, nyt toimi TäTin puolella oikein hyvin, mutta Testi-OUTIssa valutus jää vielä odotustilaan.

**Comment by johannaraisa** [comment]
> Kiitos, nyt toimi TäTin puolella oikein hyvin, mutta Testi-OUTIssa valutus jää vielä odotustilaan.

Oliskohan nyt? Noita croneja yhdenmukaistettiin ja varmaan on siinä samassa testeiltä kommentoitu.

**Comment by AnttHeik** [comment]
> > Kiitos, nyt toimi TäTin puolella oikein hyvin, mutta Testi-OUTIssa valutus jää vielä odotustilaan.
> 
> Oliskohan nyt? Noita croneja yhdenmukaistettiin ja varmaan on siinä samassa testeiltä kommentoitu.

Joo, nyt vaikuttaisi toimivan, mutta jostain syystä valutus ei reagoi siihen, jos muokkaan tietuetta Testi-TäTissä. Vie/Tuo-nappula kyllä toimii. Testasin Melinda-tietuetta 007273726 (TÄTI: 2513659, OUTI: 2381134)

**Comment by johannaraisa** [comment]
> > > Kiitos, nyt toimi TäTin puolella oikein hyvin, mutta Testi-OUTIssa valutus jää vielä odotustilaan.
> > 
> > 
> > Oliskohan nyt? Noita croneja yhdenmukaistettiin ja varmaan on siinä samassa testeiltä kommentoitu.
> 
> Joo, nyt vaikuttaisi toimivan, mutta jostain syystä valutus ei reagoi siihen, jos muokkaan tietuetta Testi-TäTissä. Vie/Tuo-nappula kyllä toimii. Testasin Melinda-tietuetta 007273726 (TÄTI: 2513659, OUTI: 2381134)

Nyt pitäisi toimia, rajapinnan käyttäjätietoihin oli jäänyt vanhaa dataa. Jospa seuraavalla kerralla nuokin saadaan siirrettyä redusoinnissa, niin ei tarvitse joka kerta tehdä uudestaan.

**Comment by AnttHeik** [comment]
Kiitos, nyt toimi oikein. Kyselen taas Melindasta testiä.

**Comment by AnttHeik** [comment]
No nyt melkein jo päästiin eteenpäin, mutta Melindalaisilla herjaa väärästä salasanasta testi-replikoinnin puolella. Tägäsin Johanna sinut tuonne Slackin keskusteluun.

**Comment by AnttHeik** [comment]
Pistän tännekin ylös testaustuloksia Melindan kanssa. Yksittäisillä tietueilla tämä toimii oikein mainiosti, mutta osakohteellisten kanssa tulee sitten ongelmia TäTin ja paikalliskannan välillä. Periaatteessa sama ongelma kuin nyttenkin välillä valutuksessa eli kaikki osakohteet eivät ehdi yhden päivityksen matkaan.

Minttu kasasi tällaisen koonnin:

- Mergetty osakohdeperhe replikoituu lähtökohtaisesti ookoo Melindasta Tätiin.
- Valuminen Tätistä paikalliskantaan on iffympää.
- Yhdistetyn emon paikalliskantaan valuminen ottaa mukaan vain ne osakohteet, jotka ovat ehtineet replikoitua (ja indeksoitua?). Tässä aktuaalinen merge olisi varmasti nopeampi kuin käsin tehty emulointi, mutta koska osakohteet replikoituvat joka tapauksessa yksi kerrallaan, etenkin isoissa osakohdeperheissä tämä ongelma esiintynee myös oikealla mergellä.
- Siirtäjä ei löydä yhdistetyn emon vanhoja osakohteita paikalliskannasta määrävertailuun ja korvattavaksi vaan ne jäävät paikalliskannassa orvoiksi -> Tässä voisi jeesata, jos siirtäjä osaisi etsiä osakohteita myös emon 035 $z:n perusteella. Kun yhdistetty emo yrittää valua paikalliskantaan seuraavan kerran, siirto tyssää siihen, että osakohdemäärät eroavat.

"Sinänsä jos se siirtäjä löytäis ne osakohteet myös 035 $z:jen perusteella paikalliskannasta, silloin ton valumisen saattais saada toimimaan sillä että:
eka emon tallennus ei tod. näk. valuisi, koska osakohdemäärät ei täsmäis
sit lisättäis mergeen ominaisuus, että se tallentais emon vielä uudelleen kaikkien osakohteiden tallennuksen jälkeen
tää emon toinen tallennus replikoituis ja sit kun siirtäjä lähtis valuttamaan perhettä uudelleen paikalliskantaan, olis toivottavasti kaikki poikaset mukana, nyt niiden määrä täsmäis ja koko perheen siirto onnistuis"

**Comment by AnttHeik** [comment]
Osakohteellisia testitietueita olivat TäTi-ID:t 2513659 ja 2513651

**Comment by johannaraisa** [comment]
Tämä vaatii kyllä pohdintaa, jos nuo muutokset tulevat Tätiin viiveellä. Vaihtaako se merge osakohteista vain kentän arvot ja päivittää Tätissä olevan osakohteen?

**Comment by AnttHeik** [comment]
> Tämä vaatii kyllä pohdintaa, jos nuo muutokset tulevat Tätiin viiveellä. Vaihtaako se merge osakohteista vain kentän arvot ja päivittää Tätissä olevan osakohteen?

Joo, se tekee osakohteille saman kuin emotietueellekin eli laittaa yhdistettävien tietueiden vanhat ID:t 35z-kenttään ja antaa osakohteelle uuden ID:n 35a-kenttään. Osakohteet päivittyvät emon jälkeen, koska niiden pitää saada emon uusi ID 773w-kenttään ensin.
